### PR TITLE
Fail restore if any warmup job failed for volume-snapshot restores

### PR DESCRIPTION
**Pull request by michaelmdeng**
### What problem does this PR solve?

Currently restore only checks if all warmup jobs completed (success or fail) before continuing w/ later steps of restore. Instead, if any warmup jobs fail, we should fail the restore.

Observed behavior:

* Failed warmup job conditions:

```yaml
...
conditions:
  - lastProbeTime: "2024-03-10T21:05:24Z"
    lastTransitionTime: "2024-03-10T21:05:24Z"
    message: Job has reached the specified backoff limit
    reason: BackoffLimitExceeded
    status: "True"
    type: Failed
failed: 5
ready: 0
startTime: "2024-03-10T20:53:09Z"
uncountedTerminatedPods: {}
```

* Results in `WarmUpComplete` in restore

```yaml
commitTs: "448239596192667254"
conditions:
  - lastTransitionTime: "2024-03-10T20:51:49Z"
    status: "True"
    type: Scheduled
  - lastTransitionTime: "2024-03-10T20:51:57Z"
    status: "True"
    type: Running
  - lastTransitionTime: "2024-03-10T20:52:11Z"
    status: "True"
    type: VolumeComplete
  - lastTransitionTime: "2024-03-10T20:52:57Z"
    status: "True"
    type: WarmUpStarted
  - lastTransitionTime: "2024-03-10T22:02:49Z"
    status: "True"
    type: WarmUpComplete
phase: WarmUpComplete
progresses:
  - lastTransitionTime: "2024-03-10T20:52:11Z"
    progress: 100
    step: Volume Restore
timeCompleted: null
timeStarted: "2024-03-10T20:51:57Z"
```

### What is changed and how does it work?

When any warmup jobs fail, fail the entire restore. Verified this in a testing workload by artificially causing warmup to fail and observing `Restore` (and consequently, `VolumeRestore`) put into `Failed` state

### Code changes

- [x] Has Go code change
- [ ] Has CI related scripts change

### Tests

- [x] Unit test 
- [ ] E2E test 
- [ ] Manual test 
- [ ] No code 

### Side effects

- [ ] Breaking backward compatibility 
- [ ] Other side effects: 

### Related changes

- [x] Need to cherry-pick to the release branch 
- [ ] Need to update the documentation 

### Release Notes

Please refer to [Release Notes Language Style Guide](https://github.com/pingcap/tidb-operator/blob/master/docs/release-note-guide.md) before writing the release note.

```release-note

```

### CODE REVIEW

**Review by YuJuncen — APPROVED** [review]
LGTM, thank you!

File: pkg/backup/restore/restore_manager_test.go [hunk 1]
```diff
@@ -674,6 +704,148 @@ func TestVolumeNumMismatchBRRestoreByEBS(t *testing.T) {
...
+							Status: corev1.ConditionTrue,
+						},
+					},
+				},
+			},
+		},
+	}
+
+	// Verify invalid tc with mismatch tikv replicas
+	//generate the restore meta in local nfs, with 3 volumes for each tikv
```
Comment (BornChanger) [inline pkg/backup/restore/restore_manager_test.go:822]:
```suggestion
```

File: pkg/backup/restore/restore_manager_test.go [hunk 2]
```diff
@@ -674,6 +704,148 @@ func TestVolumeNumMismatchBRRestoreByEBS(t *testing.T) {
...
+	err := os.WriteFile("/tmp/restoremeta", []byte(testutils.ConstructRestoreMetaStr()), 0644) //nolint:gosec
+	g.Expect(err).To(Succeed())
+
+	//generate the backup meta in local nfs, tiflash check need backupmeta to validation
```
Comment (BornChanger) [inline pkg/backup/restore/restore_manager_test.go:826]:
```suggestion
```

**Review by BornChanger — COMMENTED** [review]
Others LGTM.

**Review by BornChanger — APPROVED** [review]

File: pkg/backup/restore/restore_manager.go [hunk 3]
```diff
@@ -1276,15 +1276,24 @@ func (rm *restoreManager) waitWarmUpJobsFinished(r *v1alpha1.Restore) error {
 	for _, job := range jobs {
 		jobFinished := false
 		for _, condition := range job.Status.Conditions {
-			if condition.Type == batchv1.JobFailed || condition.Type == batchv1.JobComplete {
+			if condition.Type == batchv1.JobFailed {
```
Comment (BornChanger) [inline pkg/backup/restore/restore_manager.go:1279]: please also change restoreManager::makeSyncWarmUpJob() function to reduce the BackoffLimit. @michaelmdeng

Comment (michaelmdeng) [reply pkg/backup/restore/restore_manager.go:1279]: @BornChanger just updated, thanks!

**Comment by nkg-** [comment]
@BornChanger : Can you review again.

**Review by WangLe1321 — APPROVED** [review]

**Pull request merged**